examples/rlhf/data/openai_summarize_comparisons.py

```diff
@@ -0,0 +1,114 @@
+
+
+class Collate(nn.Module):
+    def __init__(self, device="cpu"):
+        super().__init__()
+        self.device = torch.device(device)
+
+    def __call__(self, batches):
+        chosen_batch, rejected_batch = batches
+        if self.device.type == "cuda":
+            chosen_batch = chosen_batch.pin_memory()
+            rejected_batch = rejected_batch.pin_memory()
+        return chosen_batch.to(self.device), rejected_batch.to(self.device)
+
+
+def make_process_fn(tokenizer, max_length):
+    def process(example):
+        return tokenizer(
+            example["text"],
+            max_length=max_length,
+            padding="max_length",
+            truncation=True,
+        )
+
+    return process
+
+
+def pre_tokenization_hook(dataset):
+    chosen = []
+    rejected = []
+    for sample in tqdm(dataset):
+        prompt = sample["prompt"]
+        chosen_summary = sample["chosen"]
+        rejected_summary = sample["rejected"]
+        if chosen_summary == rejected_summary:
+            continue
+        if len(chosen_summary.split()) < 5 or len(rejected_summary.split()) < 5:
+            continue
+        chosen.append({"text": prompt + "\n" + chosen_summary})
+        rejected.append({"text": prompt + "\n" + rejected_summary})
+
+    return HFDataset.from_list(chosen + rejected)
+
+
+class PairwiseDataset(Dataset):
+    def __init__(self, split, max_length=550):
+        data_dir = HERE / "comparisons"
+        ids_filename = data_dir / f"ids-{split}-{max_length}.bin"
+        mask_filename = data_dir / f"mask-{split}-{max_length}.bin"
+
+        if not all(
+            (data_dir / file).exists() for file in (ids_filename, mask_filename)
+        ):
+            create_memmaps(
+                split, max_length, DATASET, make_process_fn, pre_tokenization_hook
+            )
+
+        self.input_ids = np.memmap(ids_filename, dtype=np.int32, mode="r+")
+        self.mask = np.memmap(mask_filename, dtype=np.int32, mode="r+")
+
+        self.input_ids = self.input_ids.reshape(
+            (self.input_ids.shape[0] // max_length, max_length)
+        )
+        self.mask = self.mask.reshape((self.mask.shape[0] // max_length, max_length))
+
+    def __len__(self):
+        return len(self.input_ids) // 2
+
+    def __getitems__(self, idx):
+        ridx = [i + self.__len__() for i in idx]
+        chosen_ids = torch.from_numpy(self.input_ids[idx])
+        rejected_ids = torch.from_numpy(self.input_ids[ridx])
+        chosen_mask = torch.from_numpy(self.mask[idx])
+        rejected_mask = torch.from_numpy(self.mask[ridx])
+        batch_size = chosen_ids.shape[0]
+        chosen_data = RewardData(
+            input_ids=chosen_ids,
+            attention_mask=chosen_mask,
+            batch_size=[batch_size],
+        )
+        rejected_data = RewardData(
+            input_ids=rejected_ids,
+            attention_mask=rejected_mask,
+            batch_size=[batch_size],
+        )
+        return chosen_data, rejected_data
```

Comment:
This seems to be doing stuff that we could do with tensordict as a dataset, directly interfacing with memmap tensors rather than numpy.
It feels strange to me to see that we rely on `torch.from_numpy(np.memmap(...))` when we have `MemmapTensor` to do it for us

Reply:
Also we create manually a RewardData at every `__getitems__` call but `RewardData` has its `__getitem__`
`PairwiseDataset` should be a tensordict/tensorclass with two fields, then we can just index it without manually writing the indexing ops